src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,726 @@
+            }
+
+            return null;
+        }
+
+        /// Where each set bit represents the index of a way that has the same tag.
+        const Ways = meta.Int(.unsigned, layout.ways);
+
+        inline fn search_tags(tags: *[layout.ways]Tag, tag: Tag) Ways {
+            const x: Vector(layout.ways, Tag) = tags.*;
+            const y: Vector(layout.ways, Tag) = @splat(layout.ways, tag);
+
+            const result: Vector(layout.ways, bool) = x == y;
+            return @ptrCast(*const Ways, &result).*;
+        }
+
+        pub fn put_no_clobber(self: *Self, key: Key) *align(value_alignment) Value {
+            return self.put_no_clobber_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Add a key, evicting an older entry if needed, and return a pointer to the value.
+        /// The key must not already be in the cache.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn put_no_clobber_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (
+                Context,
+                *align(value_alignment) const Value,
+            ) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) *align(value_alignment) Value {
+            const set = self.associate(key);
+
+            if (verify) assert(self.search(set, key) == null);
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            var way = self.clocks.get(clock_index);
+            comptime assert(math.maxInt(@TypeOf(way)) == layout.ways - 1);
+            comptime assert(@as(@TypeOf(way), math.maxInt(@TypeOf(way))) +% 1 == 0);
+
+            const clock_iterations_max = layout.ways * math.maxInt(Count);
+            var safety_count: usize = 1;
+            while (safety_count <= clock_iterations_max + 1) : ({
+                safety_count += 1;
+                way +%= 1;
+            }) {
+                // We pass a value pointer to the callback here so that a cache miss
+                // can be avoided if the caller is able to determine if the value is
+                // locked by comparing pointers directly.
+                if (locked(context, @alignCast(value_alignment, &set.values[way]))) continue;
+
+                var count = self.counts.get(set.offset + way);
+                if (count == 0) break; // Way is already free.
+
+                count -= 1;
+                self.counts.set(set.offset + way, count);
+                if (count == 0) break; // Way has become free.
+            }
+            assert(safety_count <= clock_iterations_max);
```

Comment:
```suggestion
            assert(safety_count > 0);
            assert(safety_count <= clock_iterations_max);
```

Since we expect at least one iteration, and want to be sure that we called `break`.

Reply:
Why do we expect at least one iteration? I don't think we do...

Reply:
Because if we never enter the loop, we never check if the way is actually free.

Reply:
We could make this explicit by asserting that the `count` for the way is `0` before setting the count.

Reply:
We always expect to enter the loop, but we may not complete that first iteration and increment the safety count if e.g. the set is empty.

> We could make this even harder to get wrong by asserting that the `count` for the way is `0` before setting the count.

Good idea :+1:

Reply:
Ah, of course. `assert(safety_count > 0)` would not have worked. But testing for the effect of the break, i.e. asserting that the count is 0 is better.

Reply:
Pushed a commit implementing this and the other suggestion :)